pkg/secrets/pushtofile/retrieve_secrets_test.go

```diff
@@ -0,0 +1,122 @@
+
+func valuesAreEqual(values1, values2 map[string]string) bool {
+	return reflect.DeepEqual(values1, values2)
+}
+
+func assertGoodResults(expectedGroupValues map[string]map[string]string) func(*testing.T, SecretGroups, error) {
+	return func(t *testing.T, result SecretGroups, err error) {
+
+		if !assert.NoError(t, err) {
+			return
+		}
+		for groupLabel, expValues := range expectedGroupValues {
+			actualValues := findGroupValues(result, groupLabel)
+			assert.NotNil(t, actualValues)
+			assert.True(t, valuesAreEqual(*actualValues, expValues))
+		}
+	}
+}
+
+var retrieveSecretsTestCases = []retrieveSecretsTestCase{
+	{
+		description: "Happy Case",
+		secretSpecs: map[string][]SecretSpec{
+			"cache": []SecretSpec{
+				{Alias: "api-url", Id: "dev/openshift/api-url"},
+				{Alias: "username", Id: "dev/openshift/username"},
+				{Alias: "password", Id: "dev/openshift/password"},
+			},
+			"db": []SecretSpec{
+				{Alias: "api-url", Id: "ci/openshift/api-url"},
+				{Alias: "username", Id: "ci/openshift/username"},
+				{Alias: "password", Id: "ci/openshift/password"},
+			},
+		},
+		assert: assertGoodResults(map[string]map[string]string {
+			"cache": map[string]string{
+				"api-url":  string("https://postgres.example.com"),
+				"username": string("admin"),
+				"password": string("open-$e$ame"),
+			},
+			"db": map[string]string{
+				"api-url":  string("https://ci.postgres.example.com"),
+				"username": string("administrator"),
+				"password": string("open-$e$ame"),
+			},
+		}),
+	},
+	{
+		description: "Bad ID",
+		secretSpecs: map[string][]SecretSpec{
+			"cache": []SecretSpec{
+				{Alias: "api-url", Id: "foo/openshift/bar"},
+				{Alias: "username", Id: "dev/openshift/username"},
+				{Alias: "password", Id: "dev/openshift/password"},
+			},
+			"db": []SecretSpec{
+				{Alias: "api-url", Id: "ci/openshift/api-url"},
+				{Alias: "username", Id: "ci/openshift/username"},
+				{Alias: "password", Id: "ci/openshift/password"},
+			},
+		},
+		assert: func(t *testing.T, result SecretGroups, err error) {
+			assert.Contains(t, err.Error(), "Failed to retrieve secrets")
+		},
+	},
+}
+
+func TestRetrieveSecrets(t *testing.T) {
+	var mockAccessToken conjurMocks.MockAccessToken
+	conjurMockClient := conjurMocks.NewConjurMockClient()
+	conjurMockClient.AddSecret("dev/openshift/api-url", "https://postgres.example.com")
+    conjurMockClient.AddSecret("dev/openshift/username", "admin")
+    conjurMockClient.AddSecret("dev/openshift/password", "open-$e$ame")
+    conjurMockClient.AddSecret("ci/openshift/api-url", "https://ci.postgres.example.com")
+    conjurMockClient.AddSecret("ci/openshift/username", "administrator")
+    conjurMockClient.AddSecret("ci/openshift/password", "open-$e$ame")
+
+	for _, tc := range retrieveSecretsTestCases {
+		t.Run(tc.description, func(t *testing.T) {
+			s := createSecretGroups(tc.secretSpecs)
+			err := iterateOverGroups(mockAccessToken, &s, conjurMockClient.RetrieveSecrets)
+			tc.assert(t, s, err)
+		})
```

Comment:
I have recently taken to defining a `Run` method on the test spec so that 
1. The loop is a lot simpler and in its body we just call `tc.Run(t, ...deps)`
2. Each instance of the test case becomes easier to run in isolation because it defines how it needs to be run